### DISCUSSION TITLE
Getting no channels available at the end of test cases execution.

### DISCUSSION THREAD

**Discussion by mathiramu** [rabbitmq/rabbitmq-server#2446] [Other] [answered]
**********

**Comment by lukebakken** [comment]
More than likely you have a bug in your test case code. Perhaps you are not closing channels as expected. RabbitMQ does a have a per-connection channel limit. What is logged by RabbitMQ when this happens?

If you can reliably reproduce this, please provide code so that we can do the same. Thanks.

**Comment by michaelklishin** [comment, marked as answer]
[Monitoring channel churn and totals](https://www.rabbitmq.com/channels.html#monitoring) will provide useful data points. So will [server logs](https://www.rabbitmq.com/logging.html) and a traffic capture.

This message is not logged by RabbitMQ but rather by a client, [presumably this one](https://github.com/php-amqp/php-amqp/issues/225). Our team uses [php-amqplib](https://github.com/php-amqplib/php-amqplib) in our tutorials and maintains it to some extent. We are not familiar with `php-amqp` and IIRC it is based on a C library, which makes analyzing its behavior significantly more difficult.

**Comment by mathiramu** [comment]
**************

**Comment by lukebakken** [comment]
```
Cannot declare or bind RabbitMQ queue "" to the exchange point. Driver error: AMQPException: Library error: connection closed unexpectedly in
```

Your RabbitMQ log file should show why the connection was closed unexpectedly. Please find the log lines that correspond to this.

**Comment by michaelklishin** [comment]
The log entries @lukebakken is referring to are documented in [Connection Lifecycle Events](https://www.rabbitmq.com/logging.html#logged-events).

**Comment by mathiramu** [comment]
***********

**Comment by michaelklishin** [comment]
According to these logs, your application (or library) closes connections about 20 ms after opening them, cleanly. My best guess is that this is a race condition between this "connection management" by the library and what your own code does.

Such use cases are considered to be [problematic and pathological](https://www.rabbitmq.com/connections.html#high-connection-churn): RabbitMQ and the messaging protocols it supports expect connections to be reasonably long-lived.
Consider using a [specialized proxy](https://github.com/cloudamqp/amqproxy) to avoid it since this is the only known way for PHP programming model to use RabbitMQ.

Consider using `php-amqplib`, a well known, widely used, well maintained pure PHP client instead of what you use on top of `librabbitmq-c`.

**Reply by michaelklishin** [reply]
This [post](https://www.cloudamqp.com/blog/2019-05-29-maintaining-long-lived-connections-with-AMQProxy.html) explains how clients negotiate connection settings and open a connection, and what this proxy does to drastically decrease the overhead of short-lived connections on PHP applications.
